Load `content-by-url` from Rackspace Cloud Files

iangilman opened the pull request:
Don't land yet! Needs help from @aseemk @gasi @kenperkins and/or @nein09.

iangilman commented on lib/content._coffee:
```diff
@@ -6,21 +6,18 @@ redis = require 'redis'
 Worker = require './worker'
 
+pkgcloud = require('pkgcloud')
+cloudClient = pkgcloud.storage.createClient
```
I figure we might as well only do this once up here, though I'm open to other locations, of course.

aseemk replied:
I agree this is the right thing to do.

aseemk replied:
Maybe call this `cloudFiles` or `storageClient` instead of `cloudClient`, since it's specific to files/storage.

aseemk replied:
Also, a minor code style thing: it's good to keep all the `require`s together up top (and we keep them in alphabetical order since the lists can get long), so move just the `pkgcloud = ` line to the previous block of `require`s.

(Also minor: no parens needed there.)

aseemk replied:
Btw, looks like we have this same code in `worker._coffee`, and I called it `storageClient` there.

https://github.com/zoomhub/zoomhub/blob/cloud-files/lib/worker._coffee#L39-L45

Doesn't have to be for this PR, but maybe at some point it'll make sense to de-dupe this with a util, since otherwise we'll also need to duplicate stream-handling logic etc. (which you're asking about below).

iangilman commented on lib/content._coffee:
```diff
@@ -47,11 +44,14 @@ getFilePathForId = (id) ->
+        container: 'content/content-by-url'
+        remote: "#{hashURL url}.txt"
+
+    # Somehow after the download we get the ID and then ...
```
This is where my understanding of Streamline + PkgCloud runs out. I guess cloudClient.download returns a pipe or a stream or something? I just copied the code from: 

https://github.com/pkgcloud/pkgcloud#download-a-file

...but I don't know anything about pipes or streams. At any rate, we need to get the data from the pipe/stream and use it (as `id`) in the line below.

aseemk replied:
Yep, `download` returns a [readable stream](http://nodejs.org/dist/v0.10.36/docs/api/stream.html#stream_api_for_stream_consumers). Readable streams emit `end` (success) or `error` (failure) events, with `data` events beforehand for whatever's been read.

It looks like `cloudClient.download` *also* takes a callback... but it looks like it only passes the metadata for the file, not the file itself. It supports downloading the file to disk, but we probably don't want that for these tiny `id` files.

So we have to implement the logic for piecing together the `data` events, like that stream documentation example shows. And then we'll listen for `end` and `error` to either resume or fail, like here:

https://github.com/zoomhub/zoomhub/blob/cloud-files/lib/worker._coffee#L212-L213

iangilman commented on lib/content._coffee:
```diff
@@ -47,11 +44,14 @@ getFilePathForId = (id) ->
+
+    # Somehow after the download we get the ID and then ...
+    path = Path.join DIR_BY_ID_PATH, "#{id}.json"
+    readFile path, _
```
So bypassing the _ into here the actual data we want magically appears where it's needed by whoever called readFileForURL, correct?

aseemk replied:
Yep!

iangilman commented on lib/content._coffee:
```diff
@@ -276,12 +276,8 @@ module.exports = class Content
-
             FS.writeFile idPath, content._stringifyData(), _
-            FS.symlink urlToIdPath, urlPath, _
+            # TODO: Update content-by-url in Cloud Files
```
I figure we can just leave this unfixed at least until we start accepting new URLs (or maybe not ever if we go over to redis first)...at any rate this code isn't being run at this point.

gasi replied:
:+1:

iangilman commented:
Hmm...I guess another question is once we think this might work, how do I test it?

aseemk commented on lib/content._coffee:
```diff
@@ -47,11 +44,14 @@ getFilePathForId = (id) ->
+readFileForURL = (url, _) ->
+    cloudClient.download
+        container: 'content/content-by-url'
+        remote: "#{hashURL url}.txt"
```
I believe `content-by-url` needs to be part of the path (`remote`). I think `container` is just like S3 bucket.

Might also be good to make this a constant or config, in case we want to separate production from development.

iangilman replied:
When I tested the raw call, including `content-by-url` in the `container` worked. I wouldn't be surprised if it also works to put it in the `remote`, and if it's more correct to do that I'll go ahead and change it. 

I don't think I'll add a config yet, until we figure out how we want to separate it for development.

aseemk commented:
Thanks for kicking this off, @iangilman! Hope my feedback helps for now, otherwise I might be able to carve out some time to take this work over, maybe later this weekend.

iangilman commented:
Awesome, thank you for the feedback! You're welcome to take this over; just let me know. Otherwise I'll try to make some progress on Monday.

iangilman commented:
@aseemk Ok, I've addressed your comments (and added another question). Assuming it looks good to you, the next question is how do we test to see if it works? I can `npm start` to run it locally, but when I do so and then try this urls: 

http://localhost:8000/v1/content/5eha

...which works on the live zoom.it, I get "No content with ID: 5eha". Presumably I need to put something different in my config.coffee; I'll need some help with that. 

When I go to: 

http://localhost:8000/v1/content?url=http%3A%2F%2Fwww.fnrcnewbury.org.uk%2Fimages%2Fmaps%2FFullMap_Total.tif

I get a nice big stack trace. I'm not sure if this is because my error handling isn't working yet and the config isn't set up, or if there is a real bug in the code. Of course the first thing is just to get those things sorted out.

iangilman commented on lib/content._coffee:
```diff
@@ -47,11 +44,22 @@ getFilePathForId = (id) ->
+        id += chunk
+
+    download.on 'error', () ->
+        # What do we do here? Something with Streamline?
```
I don't know enough about Streamline to know what to do here.

gasi replied:
No worries, I fixed it :+1:

gasi commented:
> @iangilman: Assuming it looks good to you, the next question is how do we test to see if it works?

`npm test` seems to cover the `/v1/content/?url=…` scenario :+1: 


> I can npm start to run it locally, but when I do so and then try this urls: http://localhost:8000/v1/content/5eha

This is because locally we only have a few files checked in, e.g. http://localhost:8000/v1/content/4rcn. See `data/content-by-id`.

> When I go to:
> 
> http://localhost:8000/v1/content?url=http%3A%2F%2Fwww.fnrcnewbury.org.uk%2Fimages%2Fmaps%2FFullMap_Total.tif
> 
> I get a nice big stack trace. I'm not sure if this is because my error handling isn't working yet and the config isn't set up, or if there is a real bug in the code. Of course the first thing is just to get those things sorted out.

That URL doesn’t work for me either because we don’t have it stored locally, but this one does: http://localhost:8000/v1/content/?url=http://media.stenaline.com/media_SE/lalandia-map-zoomit/lalandia-map.jpg

gasi commented on .travis.yml:
```diff
@@ -15,3 +15,7 @@ notifications:
+
+env:
+  global:
+    - secure: "L6vgB5AL91xvQSwsxefY4YXghIzO3ztM3S/A2BCrrGVGCf5NnIYgTZIXOKSNlr2S+UYQMpPhJbvEp2DwTl68hZAquRpS7wB1ad+zQvtFTRgsjeWWIWpLEpOvOHuXROLeEofKQ962aOsPWU91S+3JC8xlpmlo7qjreO8+/ghZfOk="
```
Encrypted `RACKSPACE_API_KEY`.

iangilman commented:
@gasi Thank you for taking this over! All your changes look good to me :)
